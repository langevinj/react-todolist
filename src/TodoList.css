.Todolist{
    list-style: none;
}

.mainbody{
    text-align: center;
}
