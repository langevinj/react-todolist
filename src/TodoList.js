import React, { useState } from 'react'
import './TodoList.css'
import NewTodoForm from './NewTodoForm'
import Todo from './Todo'
import EditTodoForm from './EditTodoForm'

let loadedTodos =[];

//set the todos from local storage
window.onload = (event) => {
    loadedTodos = JSON.parse(window.localStorage.getItem('tasks'));
}

const TodoList = () => {
    //load the current todos
    const [todos, setTodos] = useState(JSON.parse(window.localStorage.getItem('tasks')))

    const add = todoObj => {
        setTodos(todos => [...todos, todoObj]);
    };

    const remove = id => {
        setTodos(todos => todos.filter(todo => todo.id !== id));
    };

    const edit = (targetTodo) => (
        setTodos(todos.map(todo => editCorrect(todo, targetTodo))
    ));

    const check = id => {
        setTodos(todos.map(todo => checkIfComplete(todo, id)))
    }

    //check a newly completed todo
    function checkIfComplete(todo, id){
        if(todo.id === id){
            let copyTodo = Object.assign({}, todo)
            copyTodo.completed = true
            return copyTodo
        }
        return todo
    }

    //togglethe edit form visibility
    function toggleEditForm(evt) {
        if(evt.target.parentNode.className === "Todo-item"){
            if (evt.target.parentNode.lastChild.style.display === "none") {
                evt.target.parentNode.lastChild.style.display = "block"
            } else {
                evt.target.parentNode.lastChild.style.display = "none"
            }
        } else {
            if (evt.target.parentNode.style.display === "none") {
                evt.target.parentNode.style.display = "block"
            } else {
                evt.target.parentNode.style.display = "none"
            }
        }
    }

    //edit the targetted todo
    function editCorrect (todo, targetTodo){
        if(todo.id === targetTodo.id){
            let copyTodo = Object.assign({}, todo)
            copyTodo.task = targetTodo.task
            return copyTodo
        }
        return todo
    }

    const todoComponents = todos.map(todo => (
        <Todo id={todo.id} task={todo.task} handleRemove={remove} handleToggle={toggleEditForm} handleEdit={edit} key={todo.id} handleCheck={check}  completed={todo.completed}/>
    ));
    
    //before page refresh, save the current todos to local storage
    window.onbeforeunload = () => {
        window.localStorage.setItem('tasks', JSON.stringify(todos))
    }

    return (
        <div className="mainbody">
            <NewTodoForm createTodo={add}/>
            <div>
                <h2 className="TodoList-title">My Todo List:</h2>
            </div>
            <ul className="Todolist">
                {todoComponents}
            </ul>
        </div>
    )
}

export default TodoList;