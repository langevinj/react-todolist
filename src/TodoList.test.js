import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList'

function addTodo(todos, task="Write Tests") {
    const taskInput = todos.getByLabelText("New Task:");
    fireEvent.change(taskInput, { target: { value: task}});
    const button = todos.getByText("Add to the todo list!");
    fireEvent.click(button);
}

it("renders without crashing", function() {
    render(<TodoList />)
});

it("matches snapshot", function() {
    const { asFragment } = render(<TodoList />)
    expect(asFragment()).toMatchSnapshot();
});

it("can add a new todo", function (){
    const todos = render(<TodoList />);

    //no todos on the list yet
    expect(todos.queryByTestId("remove-todo")).not.toBeInTheDocument();

    addTodo(todos);

    expect(todos.getByTestId("remove-todo")).toBeInTheDocument();

    //expect for to be empty
    expect(todos.getAllByDisplayValue("")).toHaveLength(1);
});

it("can remove a todo", function() {
    const todos = render(<TodoList />);
    addTodo(todos);

    const removeButtom = todos.getByTestId("remove-todo");

    fireEvent.click(removeButtom);
    expect(removeButtom).not.toBeInTheDocument();
})