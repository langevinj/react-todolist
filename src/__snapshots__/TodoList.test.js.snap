// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mainbody"
  >
    <div>
      <form
        class="NewTodoForm"
      >
        <div>
          <label
            class="NewTaskLabel"
            for="task"
          >
            New Task:
          </label>
          <textarea
            cols="20"
            id="task"
            name="task"
            rows="5"
            type="text"
          />
        </div>
        <button
          id="newTodoButton"
        >
          Add to the todo list!
        </button>
      </form>
    </div>
    <div>
      <h2
        class="TodoList-title"
      >
        My Todo List:
      </h2>
    </div>
    <ul
      class="Todolist"
    />
  </div>
</DocumentFragment>
`;
